operator/v1/types_machineconfiguration.go

```diff
@@ -56,3 +64,94 @@ type MachineConfigurationList struct {
 	// Items contains the items
 	Items []MachineConfiguration `json:"items"`
 }
+
+type ManagedBootImages struct {
+	// machineManagers is an array of machineManager objects.
+	// The MCO will watch for changes to this list and enroll/un-enroll machine management resources. An entry in this list
+	// consists of the resource type, the API group that the resource belongs to and a selection filter on the resources.
+	// All fields are required.
+	//
+	// Warning: Only one entry per machine management resource type is allowed. If multiple entries target the same type of machine management
+	// resource, all entries associated with that resource type will be ignored. For example, if one entry describes that MAPI machinesets are
+	// disabled for boot image updates and a later entry describes that MAPI machinesets are enabled for boot image updates, both entries
+	// will be ignored.
```

Comment:
Hmm, I wonder if there is a way we can set a priority, so that we don't have to have this caveat 🤔 
What if I want all MachineSets in a namespace, but, not a particular machineset? I guess I can achieve that with the matchexpressions?

Reply:
Yeah, I figure a match expression would work for that case. Couple of other approaches to avoid this caveat I had was:
- Only respect the last entry for a machine management resource if there are duplicates. This seems rather unfriendly to me but atleast we are respecting some of the user's needs
- Enforce a unique resource/apigroup pair. Two ways to do this:
-- Instead of a list of machineManagers with unbound size, have three specific definitions of machineManagers for MAPI MachineSets, CAPI MachineSets and CAPI MachineDeployments. We can add support for more types if we see this changing:
```
MAPIMachineSetManager MachineManager `json:"mapiMachineManager"`
MAPIMachineSetManager MachineManager `json:"mapiMachineManager"`
```
--

Reply:
Personally, I think we want to keep the machineManagers as it is, but add
```
// +listType=map
// +listMapKey=apiGroup
// +listMapKey=resource
```

With that, we restrict each item in the list, there must be only one entry for each combination of `apiGroup` and `resource`, which currently, limits us to 1. But expands as we add more values in the future to the enums.

Then, the partial selector becomes the only way to selector or deselect machinesets. You could use `matchExpressions` to be expressive about which to include or not include.

This is 2 in your suggestion.

I prefer 2 over 1 because it means we have specific fields which are hard to name, and, it prevents future expansion into having a namespace field.

FYI, if we add `namespace` in the future, that would also have to be a `listMapKey`.

Reply:
This makes sense to me, let's go with your approach for (2). 

Thanks for the example - I totally missed that `listMapKey` could be used in that fashion.